### PR TITLE
WIP: Remove brackets([]) from URL building in order for new modules t…

### PR DESCRIPTION
**Pull request by jmcgill298**
…o work properly

### CODE REVIEW

**Comment by jmcgill298** [comment]
@dagwieers @schunduri I still need to write integration tests for two of the modules before this is ready to merge, but please look over the current commits and let me know if there are any concerns.

The main goal of this is to move the required brackets to modules that need to use them; the brackets are required when the MO identifier uses special characters ("/" being the most common). The current method of using these for all MOs does not work for vlan/vsan pools that require only part of the MO to be in brackets.

In order to consider the changes safe, I want to add integration tests for the modules that do require brackets in the URL.

Since I was writing integration tests, I thought it made sense to fix the aci_bd_subnet bug where `scope` can have multiple values(https://github.com/datacenter/aci-ansible/issues/162). I have integrated Dag's changes and adjusted per the notes I made in that PR (https://github.com/datacenter/aci-ansible/pull/164)

File: library/aci_bd_subnet.py [hunk 1]
```diff
@@ -98,7 +104,97 @@
     aliases: [ tenant_name ]
 '''
 
-EXAMPLES = r''' # '''
+EXAMPLES = r'''
+- name: CREATE A TENANT
```
Comment (dagwieers) [inline library/aci_bd_subnet.py]: Can we not add uppercase titles ? :-)

File: library/aci_bd_subnet.py [hunk 2]
```diff
@@ -98,7 +104,97 @@
...
+- name: UPDATE A SUBNETS SCOPE TO PRIVATE AND SHARED
+  aci_bd_subnet:
+    hostname: apic
+    username: admin
+    password: SomeSecretPassword
+    tenant: production
+    bd: database
+    gateway: 10.1.1.1
+    mask: 24
+    scope: [private shared]
```
Comment (dagwieers) [inline library/aci_bd_subnet.py]: The example shows this as a list with 2 items, here it is a list with a single item.

**Comment by dagwieers** [comment]
Wrt. the combination of shared/private and share/public, I am not convinced this is a good solution.

And also wrt. the brackets, I think it's somewhat clumsy and would prefer a nicer interface. I think we need to reconsider the URL-construction as well as the global collection of objects in the library. If anything, I think a module should provide this information (or at least override it).

I don't disagree it would be nice if those definitions would be shared between modules that need the same definitions, but this solutions is worse than what it is solving IMO. I need to think about an alternative/preferred implementation.

File: library/aci_bd_subnet.py (hunk 1 above)
Comment (jmcgill298) [reply library/aci_bd_subnet.py]: sorry, that is a work standard that has become natural habit; will update

File: library/aci_bd_subnet.py (hunk 2 above)
Comment (jmcgill298) [reply library/aci_bd_subnet.py]: ah, looks like I missed a comma, will update

**Comment by jmcgill298** [comment]
```
Wrt. the combination of shared/private and share/public, I am not convinced this is a good solution.
```
1) I do not think we should require the users to construct the proper string for the module to build a correct API body, or for the module to be idempotent (the order matters)
2) The solution is simple for users to follow
3) The solution gives proper error messages when incorrect data is provided
4) The majority use case is simple public or private, which can be passed as a string in the task and Ansible converts to a list of 1 item.

If you want a different solution, then I am will to change, but I cannot budge on the module building the correct string to send in the API body; that needs to be abstracted from the user.

**Comment by jmcgill298** [comment]
```
And also wrt. the brackets, I think it's somewhat clumsy and would prefer a nicer interface. I think we need to reconsider the URL-construction as well as the global collection of objects in the library. If anything, I think a module should provide this information (or at least override it). (see #150 and #151)

I don't contest it would be nice if those definitions would be shared between modules that need the same definitions, but this solutions is worse than what it is solving IMO. I need to think about an alternative/preferred implementation. (especially given that some objects have different potential URLs that only the module can drive properly IMO)
```
I agree it it clumsy, but has other benefits. This could probably be cleaned up by making every module build each necessary dictionary for the module and adding additional keys to the dictionaries, but I need to validate this would work, and then it is one clumsy implementation that is universal, verses a less clumsy implementation that is more complex.

**Pull request closed without merging**